### ISSUE TITLE
Add numpy.r_[...] trick (RClass)?

### ISSUE THREAD

**Issue by CallumJHays** [v923z/micropython-ulab#278]
https://numpy.org/doc/stable/reference/generated/numpy.r_.html

I've managed to contort the original numpy code to get this (mostly) working in micropython, but I'm wondering if this would be worth providing a c implementation? (the numpy implementation is in python). Alternatively, would it be worth it to package a python module with this functionality, or just put it somewhere in the docs incase someone needs it?

Here's my current implementation:
```python
from typing import List, Tuple, Union  # upip.install("pycopy-typing")
from ulab import array, arange, concatenate, linspace

ndarray = array
_DType = int
RClassKeyType = Union[slice, int, float]

# this is a stripped down version of RClass (used by np.r_[...etc])
# it doesn't include support for string arguments as the first index element
class RClass:

    def __getitem__(self, key: Union[RClassKeyType, Tuple[RClassKeyType, ...]]):

        if not isinstance(key, tuple):
            key = (key,)

        objs: List[ndarray] = []
        scalars: List[int] = []
        arraytypes: List[_DType] = []
        scalartypes: List[_DType] = []

        # these may get overridden in following loop
        axis = 0

        for idx, item in enumerate(key):
            scalar = False
            if isinstance(item, slice):
                step = item.step
                start = item.start
                stop = item.stop
                if start is None:
                    start = 0
                if step is None:
                    nstep = 1
                if isinstance(step, complex):
                    size = int(abs(step))
                    newobj = cast(ndarray, linspace(start, stop, num=size))
                else:
                    newobj = arange(start, stop, step)

            # if is number
            elif isinstance(item, (int, float)):
                newobj = array([item])
                scalars.append(len(objs))
                scalar = True
                scalartypes.append(newobj.dtype())

            else:
                raise Exception("index object %s of type %s is not supported by r_[]" % (
                    str(item), type(item)))

            objs.append(newobj)
            if not scalar and isinstance(newobj, ndarray):
                arraytypes.append(newobj.dtype())

        # Ensure that scalars won't up-cast unless warranted
        # TODO: ensure that this actually works for dtype coercion
        # likelihood is we're going to have to do some funky logic for this
        final_dtype = max(arraytypes + scalartypes)
        if final_dtype is not None:
            for idx in scalars:
                objs[idx] = array(objs[idx], dtype=final_dtype)

        res = concatenate(tuple(objs), axis=axis)

        return res

    # this seems weird - not sure what it's for
    def __len__(self):
        return 0


r_ = RClass()
```

EDIT: Get example working standalone
EDIT 2: Re-add support for complex slice `step` (for linspace)

**Comment by v923z** [comment]
@CallumJHays I thought that I commented on this issue, but obviously, this is not the case. I would definitely like you to move this piece of code to a more visible place. I was thinking that we could create a `snippets` directory for functions that are implemented in `python`, and this definitely belongs there.

**Comment by v923z** [comment]
@CallumJHays Cal, is this OK, or do you want to change something here: https://github.com/v923z/micropython-ulab/pull/358?

**Comment by CallumJHays** [comment]
Merged in with #358
